bridge/server/services/data-service.ts

```diff
@@ -53,87 +60,218 @@ export class DataService {
-              remediations,
-              lastSequences[service.serviceName]
-            );
-            if (latestSequence) {
-              lastSequences[service.serviceName] = latestSequence;
-            }
-          } catch (error) {
-            console.error(error);
+        const latestEvent = service.getLatestEvent();
+        if (latestEvent) {
+          const latestSequence = await this.fetchServiceDetails(
+            service,
+            stage.stageName,
+            latestEvent.keptnContext,
+            projectName,
+            stageInformation,
+            stageDeployments,
+            cachedSequences[latestEvent.keptnContext]
+          );
+          if (latestSequence) {
+            cachedSequences[latestEvent.keptnContext] = latestSequence;
           }
         }
       }
     }
     return project;
   }
 
+  private async getSequencesForServices(projectName: string, services: Service[]): Promise<Sequence[]> {
+    const chunkSize = 100;
+    let sequences: Sequence[] = [];
+
+    for (let i = 0; i < services.length; i += chunkSize) {
+      const keptnContexts: string[] = [];
+      const maxLength = Math.min(i + chunkSize, services.length);
+
+      for (let y = i; y < maxLength; ++y) {
+        const latestServiceEvent = services[y].getLatestEvent();
+        if (latestServiceEvent) {
+          // string concatenation is expensive; that's why we use an array here
+          keptnContexts.push(latestServiceEvent.keptnContext);
+        }
+      }
+      if (keptnContexts.length) {
+        const response = await this.apiService.getSequences(
+          projectName,
+          chunkSize,
+          undefined,
+          undefined,
+          undefined,
+          undefined,
+          keptnContexts.join(',')
+        );
+        sequences = [...sequences, ...response.data.states];
+      }
+    }
+    return sequences;
+  }
+
+  private async getDeploymentFinishedForServices(
+    projectName: string,
+    services: Service[],
+    resultType?: ResultTypes
+  ): Promise<Trace[]> {
+    return this.getTracesOfMultipleServices(
+      projectName,
+      services,
+      EventTypes.DEPLOYMENT_FINISHED,
+      (service: Service) => service.lastEventTypes[EventTypes.DEPLOYMENT_FINISHED],
+      undefined,
+      resultType
+    );
+  }
+
+  private async getEvaluationResultsForServices(
+    projectName: string,
+    services: Service[],
+    resultType?: ResultTypes
+  ): Promise<Trace[]> {
+    return this.getTracesOfMultipleServices(
+      projectName,
+      services,
+      EventTypes.EVALUATION_FINISHED,
+      (service: Service) => {
+        const latestEvent = service.getLatestEvent();
+        const latestEvaluation = service.lastEventTypes[EventTypes.EVALUATION_FINISHED];
+        return latestEvent?.keptnContext === latestEvaluation?.keptnContext ? latestEvaluation : undefined;
+      },
+      KeptnService.LIGHTHOUSE_SERVICE,
+      resultType
+    );
+  }
+
+  private async getTracesOfMultipleServices(
```

Comment:
I would also rename them to some sort of latest to not need to know implementation details